pkg/controller/sparkapplication/sparkapp_metrics.go

```diff
@@ -196,6 +218,18 @@ func (sm *sparkAppMetrics) exportMetrics(oldApp, newApp *v1beta2.SparkApplicatio
 		}
 	}
 
+	// In the event that state transitions happened too quickly and the spark app skipped the RUNNING state, the job
```

Comment:
Do we really need this? Generally Kubernetes controller should not be programmed against a state machine. I know we are already doing this in the job state handling logic, but putting in more logic that works against a state machine makes the code even more fragile and harder to read. I think we should try to gradually reduce reliance on state transitions.

Reply:
@liyinan926  There is a concern that if the work queue is huge and it takes a long time to process the next item for a certain sparkapp after it ran spark-submit on it, the driver pod would have already gone through the `PodRunning` state and be in `PodSucceeded` or `PodFailed`, then the sparkapp state will skip `RUNNING` and transit from `SUBMITTED` to `FAILING`/`SUCCEEDING` directly. This is unlikely in normal operator, but when the queue gets piled up, the metric will lose those data.

Reply:
The problem is that in this case, there's no way you can get an accurate measure of the latency.

Reply:
Yes, but at least there would be a data point indicating that the job has already started. Even the value is just an upper bound and it's not accurate, I think it's still valuable to have the data, otherwise it might trigger a lot more false positive alerts if user sets up a pager on this metric.